# To find rows in a DataFrame we use iloc and loc

In [1]:
import pandas as pd

In [5]:
df = pd.read_csv('titanic_test.csv')

In [6]:
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN       Qv  

[891 rows x 12 columns]

In [7]:
df.iloc[0]

PassengerId                          1
Survived                             0
Pclass                               3
Name           Braund, Mr. Owen Harris
Sex                               male
Age                               22.0
SibSp                                1
Parch                                0
Ticket                       A/5 21171
Fare                              7.25
Cabin                              NaN
Embarked                             S
Name: 0, dtype: object

## if there is a key then we must use the key in loc but here there is no labels

In [8]:
df.iloc[0]

PassengerId                          1
Survived                             0
Pclass                               3
Name           Braund, Mr. Owen Harris
Sex                               male
Age                               22.0
SibSp                                1
Parch                                0
Ticket                       A/5 21171
Fare                              7.25
Cabin                              NaN
Embarked                             S
Name: 0, dtype: object

In [9]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [10]:
df.loc[[0]]

PassengerId  Survived  Pclass                     Name   Sex   Age  SibSp  \
0            1         0       3  Braund, Mr. Owen Harris  male  22.0      1   

   Parch     Ticket  Fare Cabin Embarked  
0      0  A/5 21171  7.25   NaN        S

# Data Filtering

In [12]:
df[df.Age>20]

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
884          885         0       3   
885          886         0       3   
886          887         0       2   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
884                             Sutehall, Mr. Henry Jr    male  25.0      0   
885               Rice, Mrs. William (Margaret Norton)  female  39.0      0   
886                              Montvila, Rev. Juozas    male  27.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
884      0   SOTON/OQ 392076   7.0500   NaN        S  
885      5            382652  29.1250   NaN        Q  
886      0            211536  13.0000   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN       Qv  

[535 rows x 12 columns]

In [13]:
df[df.Survived == 1]

PassengerId  Survived  Pclass  \
1              2         1       1   
2              3         1       3   
3              4         1       1   
8              9         1       3   
9             10         1       2   
..           ...       ...     ...   
875          876         1       3   
879          880         1       1   
880          881         1       2   
887          888         1       1   
889          890         1       1   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
8    Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                  Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
..                                                 ...     ...   ...    ...   
875                   Najib, Miss. Adele Kiamie "Jane"  female  15.0      0   
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   
880       Shelley, Mrs. William (Imanita Parrish Hall)  female  25.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   

     Parch            Ticket     Fare Cabin Embarked  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
8        2            347742  11.1333   NaN        S  
9        0            237736  30.0708   NaN        C  
..     ...               ...      ...   ...      ...  
875      0              2667   7.2250   NaN        C  
879      1             11767  83.1583   C50        C  
880      1            230433  26.0000   NaN        S  
887      0            112053  30.0000   B42        S  
889      0            111369  30.0000  C148        C  

[342 rows x 12 columns]

In [15]:
df[df.Survived == 1].Survived.count()

342

In [18]:
df[df.Sex == 'male'].shape

(577, 12)

In [22]:
df[df.Sex == 'female'].shape

(314, 12)

# EDA (Exploratory Data Analysis)

## EDA is done because:
#### Data Acquisition
#### Data isn't always clean, in raw format

In [20]:
df[(df.Sex == 'male') & (df.Age > 20)]

PassengerId  Survived  Pclass                           Name   Sex   Age  \
0              1         0       3        Braund, Mr. Owen Harris  male  22.0   
4              5         0       3       Allen, Mr. William Henry  male  35.0   
6              7         0       1        McCarthy, Mr. Timothy J  male  54.0   
13            14         0       3    Andersson, Mr. Anders Johan  male  39.0   
20            21         0       2           Fynney, Mr. Joseph J  male  35.0   
..           ...       ...     ...                            ...   ...   ...   
883          884         0       2  Banfield, Mr. Frederick James  male  28.0   
884          885         0       3         Sutehall, Mr. Henry Jr  male  25.0   
886          887         0       2          Montvila, Rev. Juozas  male  27.0   
889          890         1       1          Behr, Mr. Karl Howell  male  26.0   
890          891         0       3            Dooley, Mr. Patrick  male  32.0   

     SibSp  Parch            Ticket     Fare Cabin Embarked  
0        1      0         A/5 21171   7.2500   NaN        S  
4        0      0            373450   8.0500   NaN        S  
6        0      0             17463  51.8625   E46        S  
13       1      5            347082  31.2750   NaN        S  
20       0      0            239865  26.0000   NaN        S  
..     ...    ...               ...      ...   ...      ...  
883      0      0  C.A./SOTON 34068  10.5000   NaN        S  
884      0      0   SOTON/OQ 392076   7.0500   NaN        S  
886      0      0            211536  13.0000   NaN        S  
889      0      0            111369  30.0000  C148        C  
890      0      0            370376   7.7500   NaN       Qv  

[351 rows x 12 columns]

## Data Wrangling or Munging

#### 1. Discovering  
#### 2. Structuring
#### 3. Cleaning
#### 4. Enriching 
#### 5. Validating 
#### 6. Publishing

Here the EDA analysis is done by first we find male with age greater than 20 and then the age is taken out which comes in a pandas Series format

Then the Series format has a count() method so we use acess the Age then count the number of male with age above 20 

In [23]:
df[(df.Sex == 'male') & (df.Age > 20)].Age.count()

351

EDA is done: graph plotting and seeing the nature of data so that we can choose the right machine learning model

We represent the data in pie chart, bar graph, histogram

In [25]:
df.Embarked.unique()

array(['S', 'C', 'Q', nan, 'Qv'], dtype=object)

In [26]:
df.Sex.unique()

array(['male', 'female'], dtype=object)

In [28]:
df[df.Age < 1]

PassengerId  Survived  Pclass                             Name     Sex  \
78            79         1       2    Caldwell, Master. Alden Gates    male   
305          306         1       1   Allison, Master. Hudson Trevor    male   
469          470         1       3    Baclini, Miss. Helene Barbara  female   
644          645         1       3           Baclini, Miss. Eugenie  female   
755          756         1       2        Hamalainen, Master. Viljo    male   
803          804         1       3  Thomas, Master. Assad Alexander    male   
831          832         1       2  Richards, Master. George Sibley    male   

      Age  SibSp  Parch  Ticket      Fare    Cabin Embarked  
78   0.83      0      2  248738   29.0000      NaN        S  
305  0.92      1      2  113781  151.5500  C22 C26        S  
469  0.75      2      1    2666   19.2583      NaN        C  
644  0.75      2      1    2666   19.2583      NaN        C  
755  0.67      1      1  250649   14.5000      NaN        S  
803  0.42      0      1    2625    8.5167      NaN        C  
831  0.83      1      1   29106   18.7500      NaN        S

In [29]:
df.Survived.dtype

dtype('int64')

In [32]:
df.Survived.max()

1

In [34]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


memory usage: 83.7+ KB is this but after we convert the dtype of Survived we get 

In [36]:
df.Survived = df.Survived.astype('int8')

In [37]:
df.Survived.dtype

dtype('int8')

In [38]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int8   
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(4), int8(1), object(5)
memory usage: 77.6+ KB


the memory usages is memory usage: 77.6+ KB

In [39]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [40]:
df.PassengerId.dtype

dtype('int64')

In [43]:
df.PassengerId = df.PassengerId.astype('int8')

In [44]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int8   
 1   Survived     891 non-null    int8   
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(3), int8(2), object(5)
memory usage: 71.5+ KB


memory usage: 71.5+ KB

In [46]:
df.Pclass.max()

3

In [47]:
 df.SibSp.max()

8

In [49]:
df.SibSp.astype('int8').info()

<class 'pandas.core.series.Series'>
RangeIndex: 891 entries, 0 to 890
Series name: SibSp
Non-Null Count  Dtype
--------------  -----
891 non-null    int8 
dtypes: int8(1)
memory usage: 1019.0 bytes


In [50]:
df.Parch = df.Parch.astype('int8')

In [52]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int8   
 1   Survived     891 non-null    int8   
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int8   
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(2), int8(3), object(5)
memory usage: 65.4+ KB


In [53]:
import numpy as np

In [55]:
df[(df.Embarked == 'S') | (df.Embarked == 'C')] 

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
884          117         0       3   
886          119         0       2   
887          120         1       1   
888          121         0       3   
889          122         1       1   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
884                             Sutehall, Mr. Henry Jr    male  25.0      0   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
884      0   SOTON/OQ 392076   7.0500   NaN        S  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  

[812 rows x 12 columns]

In [56]:
df[(df.Embarked == 'S') | (df.Embarked == 'C')].Embarked.count() 

812

isin() is used as a alternative for | (or)

In [59]:
df[df.Embarked.isin(['S','C'])]

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
884          117         0       3   
886          119         0       2   
887          120         1       1   
888          121         0       3   
889          122         1       1   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
884                             Sutehall, Mr. Henry Jr    male  25.0      0   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
884      0   SOTON/OQ 392076   7.0500   NaN        S  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  

[812 rows x 12 columns]

In [60]:
df[df.Embarked.isin(['S','C'])].Embarked.count()

812

In [61]:
df.Age.isnull().sum()

177

In [62]:
df.Age.mean()

29.69911764705882

In [63]:
df.Age.isnull().sum()

177

In [64]:
df[df.Survived == 1].count()

PassengerId    342
Survived       342
Pclass         342
Name           342
Sex            342
Age            290
SibSp          342
Parch          342
Ticket         342
Fare           342
Cabin          136
Embarked       340
dtype: int64

In [65]:
df.Name.dtype

dtype('O')

In [66]:
df.Name.str.contains('Dr').sum()

11

In [70]:
df[df.Name.str.contains('Dr.')]

PassengerId  Survived  Pclass  \
47            48         1       3   
130         -125         0       3   
245          -10         0       1   
317           62         0       2   
398         -113         0       2   
416          -95         1       2   
632          121         1       1   
660         -107         1       1   
679          -88         1       1   
766           -1         0       1   
796           29         1       1   

                                                Name     Sex   Age  SibSp  \
47                         O'Driscoll, Miss. Bridget  female   NaN      0   
130                             Drazenoic, Mr. Jozef    male  33.0      0   
245                      Minahan, Dr. William Edward    male  44.0      2   
317                             Moraweck, Dr. Ernest    male  54.0      0   
398                                 Pain, Dr. Alfred    male  23.0      0   
416  Drew, Mrs. James Vivian (Lulu Thorne Christian)  female  34.0      1   
632                        Stahelin-Maeglin, Dr. Max    male  32.0      0   
660                    Frauenthal, Dr. Henry William    male  50.0      2   
679               Cardeza, Mr. Thomas Drake Martinez    male  36.0      0   
766                        Brewe, Dr. Arthur Jackson    male   NaN      0   
796                      Leader, Dr. Alice (Farnham)  female  49.0      0   

     Parch    Ticket      Fare        Cabin Embarked  
47       0     14311    7.7500          NaN        Q  
130      0    349241    7.8958          NaN        C  
245      0     19928   90.0000          C78        Q  
317      0     29011   14.0000          NaN        S  
398      0    244278   10.5000          NaN        S  
416      1     28220   32.5000          NaN        S  
632      0     13214   30.5000          B50        C  
660      0  PC 17611  133.6500          NaN        S  
679      1  PC 17755  512.3292  B51 B53 B55        C  
766      0    112379   39.6000          NaN        C  
796      0     17465   25.9292          D17        S